### PR TITLE
global sort: speed up 2-level iterator

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #49717

Problem Summary:

### What changed and how does it work?

Currently when mergeIter drains a reader, limitSizeMergeIter opens next reader. Opening reader maybe slow if the reader needs network actions. So in this PR we open more readers in background in advance.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/iter.go [hunk 1]
```diff
@@ -619,24 +646,80 @@ func newMergePropBaseIter(
...
+		// readers we open them in advance to reduce block when we need to open them.
+		for i := int(limit); i < len(multiStat.Filenames); i++ {
+			filePair := multiStat.Filenames[i]
+			path := filePair[1]
+			asyncTask := make(chan readerAndError, 1)
+			go func() {
+				defer close(asyncTask)
+				rd, err := newStatsReader(ctx, exStorage, path, 500*1024)
+				select {
+				case <-closeCh:
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/iter.go:670]: close rd?

File: br/pkg/lightning/backend/external/iter.go [hunk 2]
```diff
@@ -619,24 +646,80 @@ func newMergePropBaseIter(
...
+					return
+				case asyncTask <- readerAndError{r: &statReaderProxy{p: path, r: rd}, err: err}:
+				}
+			}()
+			select {
+			case <-closeCh:
```
Comment (D3Hunter) [inline br/pkg/lightning/backend/external/iter.go:676]: drain preOpenCh and all it's item before return?

File: br/pkg/lightning/backend/external/iter.go [hunk 3]
```diff
@@ -419,11 +420,29 @@ func (i *limitSizeMergeIter[T, R]) next() (ok bool, closeReaderIdx int) {
...
+	i.weightSum -= i.weights[closeReaderIdx]
+	if err := i.tryOpenMoreReaders(); err != nil {
+		i.mergeIter.err = err
+		return false, closeReaderIdx
+	}
+
+	// we need to call next once because mergeIter doesn't use h[0] as current value,
+	// but a separate curr field
+	if mergeIterDrained && i.mergeIter.h.Len() > 0 {
+		_, ok = i.mergeIter.next()
```
[the remark this answers is not in the document]
Comment (lance6716) [reply br/pkg/lightning/backend/external/iter.go:445]: in this case mergeIter's `curr` is empty, so calling `next` once should only pop from the heap to `curr`.

File: br/pkg/lightning/backend/external/iter.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/iter.go:684]: https://github.com/pingcap/tidb/pull/49832/commits/25b9e94e31eb29cc1ba3fb0c647239b00fe8d782

**Review by D3Hunter — APPROVED** [review]

**Review by Benjamin2037 — APPROVED** [review]
LGTM

**Pull request merged**